File: backend/agent/runtime/compression/micro_compact.py
from __future__ import annotations

import copy
from typing import Any

from backend.agent.model.messages import Message, Role, ToolResultBlock, ToolUseBlock


# Whitelist of tools whose results can be compacted
COMPACTABLE_TOOLS = frozenset({
    "read_file_patch",
    "search_repo",
    "search_tests_for",
    "read_repo_manifest",
    "read_check_summary",
    "search_diff",
})

# Tools that are critical evidence sources — never compact their results
EVIDENCE_CRITICAL_TOOLS = frozenset({
    "read_repo_file",
})

# Tools that should never be compacted
EXCLUDED_TOOLS = frozenset({
    "finish_context_package",
})


def deep_copy_messages(messages: list[Message]) -> list[Message]:
    """Create a deep copy of messages for MicroCompact.

    Never mutates the original messages.
    """
    return copy.deepcopy(messages)


def is_compactable_tool_use(block: ToolUseBlock) -> bool:
    """Check if a tool use block is from a compactable tool."""
    return block.name in COMPACTABLE_TOOLS


def is_compactable_tool_result(
    block: ToolResultBlock,
    tool_use_name: str,
    is_error: bool,
    min_chars: int,
) -> bool:
    """Check if a tool result can be compacted.

    Conditions:
    - Tool is in compactable whitelist
    - Tool is not evidence-critical (never compacted)
    - Result is not an error
    - Result content is above minimum size threshold
    """
    if tool_use_name in EXCLUDED_TOOLS:
        return False
    if tool_use_name in EVIDENCE_CRITICAL_TOOLS:
        return False
    if tool_use_name not in COMPACTABLE_TOOLS:
        return False
    if is_error:
        return False
    if len(block.content) < min_chars:
        return False
    return True


def build_placeholder(tool_name: str, original_chars: int, tool_input: dict[str, Any] | None = None) -> str:
    """Build a compact placeholder for a tool result.

    Includes tool name, original size, and available input metadata.
    """
    parts = [
        f"[Compacted: {tool_name}]",
        f"Original size: {original_chars:,} chars",
    ]

    if tool_input:
        # Include relevant input metadata
        if "path" in tool_input:
            parts.append(f"File: {tool_input['path']}")
        if "query" in tool_input:
            parts.append(f"Query: {tool_input['query']}")
        if "filename" in tool_input:
            parts.append(f"File: {tool_input['filename']}")

    return " | ".join(parts)


def micro_compact_messages(
    messages: list[Message],
    recent_count: int = 3,
    min_chars: int = 1000,
) -> list[Message]:
    """Create a compacted copy of messages for model request.

    Replaces only old large successful compactable tool results with placeholders.
    Never mutates the original messages.

    Args:
        messages: Original messages to compact.
        recent_count: Number of recent tool results to keep unchanged.
        min_chars: Minimum character count to consider for compaction.

    Returns:
        Deep copy with compacted tool results (originals unchanged).
    """
    if not messages:
        return []

    # Deep copy to avoid mutation
    compacted = deep_copy_messages(messages)

    # Track tool results from end to identify recent ones
    tool_result_indices: list[tuple[int, int]] = []  # (msg_idx, block_idx)

    for i, msg in enumerate(compacted):
        if msg.role == Role.TOOL and isinstance(msg.content, list):
            for j, block in enumerate(msg.content):
                if isinstance(block, ToolResultBlock):
                    tool_result_indices.append((i, j))

    # Keep the last N tool results unchanged
    recent_result_ids: set[str] = set()
    if recent_count > 0:
        for _, (msg_idx, block_idx) in enumerate(tool_result_indices[-recent_count:]):
            block = compacted[msg_idx].content[block_idx]
            if isinstance(block, ToolResultBlock):
                recent_result_ids.add(block.tool_use_id)

    # Build a map of tool_use_id -> tool_name and tool_input
    tool_info: dict[str, tuple[str, dict[str, Any]]] = {}
    for msg in compacted:
        if msg.role == Role.ASSISTANT and isinstance(msg.content, list):
            for block in msg.content:
                if isinstance(block, ToolUseBlock):
                    tool_info[block.tool_use_id] = (block.name, block.input)

    # Compact eligible tool results
    for msg_idx, block_idx in tool_result_indices:
        msg = compacted[msg_idx]
        if not isinstance(msg.content, list):
            continue

        block = msg.content[block_idx]
        if not isinstance(block, ToolResultBlock):
            continue

        # Skip recent results
        if block.tool_use_id in recent_result_ids:
            continue

        # Get tool info
        tool_name, tool_input = tool_info.get(block.tool_use_id, ("unknown", {}))

        # Check if compactable
        is_error = block.is_error
        if not is_compactable_tool_result(block, tool_name, is_error, min_chars):
            continue

        # Replace with placeholder
        placeholder = build_placeholder(tool_name, len(block.content), tool_input)
        msg.content[block_idx] = ToolResultBlock(
            tool_use_id=block.tool_use_id,
            content=placeholder,
            is_error=False,
        )

    return compacted

File: server/src/agent/compression/micro-compact.ts
/**
 * Micro-compact — replace old large tool results with placeholders.
 *
 * Only targets compactable tools (not evidence-critical), keeps recent
 * results unchanged, never mutates the original message array.
 */

import type { LLMMessage, ToolUseBlock, ToolResultBlock, ContentBlock } from '@open-multi-agent/core'
// ContentBlock is used in the type assertion below

/** Tools whose results can be compacted. */
const COMPACTABLE_TOOLS = new Set([
  'read_file_patch',
  'search_repo',
  'search_tests_for',
  'read_repo_manifest',
  'read_check_summary',
  'search_diff',
])

/** Tools that are critical evidence sources — never compact. */
const EVIDENCE_CRITICAL_TOOLS = new Set(['read_repo_file'])

/** Tools that should never be compacted. */
const EXCLUDED_TOOLS = new Set(['finish_context_package', 'todo_write'])

/**
 * Deep clone messages for safe mutation.
 */
function cloneMessages(messages: LLMMessage[]): LLMMessage[] {
  // Structured clone handles the readonly arrays
  return messages.map(msg => ({
    ...msg,
    content: Array.isArray(msg.content)
      ? msg.content.map(block => ({ ...block }))
      : msg.content,
  }))
}

interface ToolInfo {
  name: string
  input: Record<string, unknown>
}

function buildPlaceholder(toolName: string, originalChars: number, toolInput?: Record<string, unknown>): string {
  const parts = [`[Compacted: ${toolName}]`, `Original size: ${originalChars.toLocaleString()} chars`]

  if (toolInput) {
    if (typeof toolInput.path === 'string') parts.push(`File: ${toolInput.path}`)
    if (typeof toolInput.query === 'string') parts.push(`Query: ${toolInput.query}`)
    if (typeof toolInput.filename === 'string') parts.push(`File: ${toolInput.filename}`)
  }

  return parts.join(' | ')
}

/**
 * Create a compacted copy of messages for the model request.
 *
 * Replaces only old, large, successful, compactable tool results with
 * placeholders.  Never mutates the original messages.
 *
 * @param messages   Original messages.
 * @param recentCount  Number of recent tool results to keep unchanged.
 * @param minChars     Minimum char count to consider for compaction.
 */
export function microCompactMessages(
  messages: LLMMessage[],
  recentCount = 3,
  minChars = 1000,
): LLMMessage[] {
  if (!messages.length) return []

  const compacted = cloneMessages(messages)

  // Collect all tool_result positions: [msgIdx, blockIdx]
  const toolResultPositions: Array<[number, number]> = []
  for (let i = 0; i < compacted.length; i++) {
    const msg = compacted[i]
    if (!Array.isArray(msg.content)) continue
    for (let j = 0; j < msg.content.length; j++) {
      if (msg.content[j].type === 'tool_result') {
        toolResultPositions.push([i, j])
      }
    }
  }

  // Keep last N tool results unchanged
  const recentResultIds = new Set<string>()
  if (recentCount > 0) {
    const recentPositions = toolResultPositions.slice(-recentCount)
    for (const [msgIdx, blockIdx] of recentPositions) {
      const block = compacted[msgIdx].content[blockIdx] as ToolResultBlock
      recentResultIds.add(block.tool_use_id)
    }
  }

  // Build map: tool_use_id -> { name, input }
  const toolInfo = new Map<string, ToolInfo>()
  for (const msg of compacted) {
    if (!Array.isArray(msg.content)) continue
    for (const block of msg.content) {
      if (block.type === 'tool_use') {
        const tb = block as ToolUseBlock
        toolInfo.set(tb.id, { name: tb.name, input: tb.input })
      }
    }
  }

  // Compact eligible tool results
  for (const [msgIdx, blockIdx] of toolResultPositions) {
    const msg = compacted[msgIdx]
    if (!Array.isArray(msg.content)) continue

    const block = msg.content[blockIdx] as ToolResultBlock

    // Skip recent results
    if (recentResultIds.has(block.tool_use_id)) continue

    // Get tool info
    const info = toolInfo.get(block.tool_use_id) ?? { name: 'unknown', input: {} }

    // Check compactability
    if (EXCLUDED_TOOLS.has(info.name)) continue
    if (EVIDENCE_CRITICAL_TOOLS.has(info.name)) continue
    if (!COMPACTABLE_TOOLS.has(info.name)) continue
    if (block.is_error) continue
    if (block.content.length < minChars) continue

    // Replace with placeholder — create new message since content is readonly
    const placeholder = buildPlaceholder(info.name, block.content.length, info.input)
    const newContent = msg.content.map((b, idx) =>
      idx === blockIdx
        ? { type: 'tool_result' as const, tool_use_id: block.tool_use_id, content: placeholder }
        : b,
    ) as unknown as ContentBlock[]
    compacted[msgIdx] = { ...msg, content: newContent }
  }

  return compacted
}
